aiven/client/cli.py

```diff
@@ -3382,6 +3431,95 @@ def _get_service_type_user_config_schema(self, project, service_type):
+        if service_type in {"m3db", "m3aggregator", "m3coordinator"}:
+            if "m3_version" in user_config:
+                if service_version_key in user_config:
+                    raise argx.UserError(f"'{service_version_key}' and 'm3_version' cannot be specified together")
+
+                # Replace m3_version with service_type specific key
+                user_config[service_version_key] = user_config.pop("m3_version")
+
+        return user_config.get(service_version_key)
+
+    def _do_version_eol_check(self, project: str, service_type: str, requested_version: str) -> None:
+        """Checks the specified service version against EOL times and rules.
+
+        EOL Rules:
+            1. If a service version is EOL, it won't be allowed
+            2. If a service has pass "availability time" it won't be allowed...
+            3. ... unless a service of that version is still exists in the project
+            4. If a service is approaching EOL, encourage user to move to new version
+        """
+        service_version = self._get_service_version_info(service_type, requested_version)
+
+        current_time = get_current_date()
+
+        if not service_version["aiven_end_of_life_time"] or not service_version["availability_end_time"]:
+            return  # No EOLs specified
+
+        end_of_life_time = parse_iso8601(service_version["aiven_end_of_life_time"])
+        availability_end_time = parse_iso8601(service_version["availability_end_time"])
+
+        warning = [
+            "   !!! WARNING !!!",
+            "",
+            f"{service_type} v{requested_version} is reaching EOL soon ({end_of_life_time.date()}).",
+            "",
+            "It is highly recommended to deploy newer, supported versions of the service.",
+        ]
+
+        if current_time > end_of_life_time:
+            # User cannot start a service already passed EOL
+            raise argx.UserError(f"{service_type} v{requested_version} has reached EOL and cannot be used")
+
+        elif current_time > availability_end_time:
+            # User can only start a service passed availability if one already exists in the project...
+            project_services = self.client.get_services(project=project)
+
+            # Check if services with the same version already exist in the project
+            services_matching_version = [
+                service for service in project_services if service["service_type"] == service_type
+                and service["user_config"][f"{service_type}_version"] == (requested_version)
+            ]
+
+            if len(services_matching_version) == 0:
+                raise argx.UserError(
+                    f"{service_type} v{requested_version} is reaching EOL and cannot be used for new services"
+                )
+
+            # ...But should be discouraged
+            self.print_boxed(warning)
+
+            if not self.confirm("continue anyway? "):
+                raise argx.UserError("Aborted")
```

Comment:
I wouldn't put these absolute checks here, as there may be cases where e.g. customer specific dispensation is made to allow for it later on and clients are updated seldom. Instead, similar checks should exist in the backend which is updated constantly.

Due to that, any condition here should be just warning with confirm prompt.

Reply:
To be honest, I wasn't fully happy with these either because, as you say, these should be in sync with the "real" rules are, as enforced by server side, which may change at any point (and considering aiven-client installations in the wild could be "old" ....)
Should we still have a difference in messaging? something like:
1)  "advance warning" -> consider deploying a newer version
2) "passed availability" -> warning, operation may fail?

or reduce it just to the 1) ?

Reply:
I don't have strong opinion either way - 2 adds some extra value, but is added complexity on this side (and something that has to be kept up to date). By default I'd prefer probably simpler choice (just 1), but I won't object to 2 either as long as it doesn't block operation.

Reply:
Went with option 1 ... KISS prevails